# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of each sentence from `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_id_text = []; target_id_text = []
    for sentence in source_text.split('\n'):
        sentence_id = []
        for word in sentence.split():
            sentence_id.append(source_vocab_to_int[word])
        source_id_text.append(sentence_id)
        
    for sentence in target_text.split('\n'):
        sentence_id = []
        for word in sentence.split():
            sentence_id.append(target_vocab_to_int[word])
        sentence_id.append(target_vocab_to_int['<EOS>'])
        target_id_text.append(sentence_id)    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf

# Check TensorFlow Version
assert LooseVersion(tf.__version__) in [LooseVersion('1.0.0'), LooseVersion('1.0.1')], 'This project requires TensorFlow version 1.0  You are using {}'.format(tf.__version__)
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.0.1
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoding_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.

Return the placeholders in the following the tuple (Input, Targets, Learing Rate, Keep Probability)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, and learning rate.
    :return: Tuple (input, targets, learning rate, keep probability)
    """
    # TODO: Implement Function
    input = tf.placeholder(tf.int32, shape=[None,None],name="input")
    targets = tf.placeholder(tf.int32, shape=[None,None],name='targets')
    learning_rate = tf.placeholder(tf.float32,name='learning_rate')
    keep_prob = tf.placeholder(tf.float32,name='keep_prob')
    return input, targets, learning_rate, keep_prob

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoding Input
Implement `process_decoding_input` using TensorFlow to remove the last word id from each batch in `target_data` and concat the GO ID to the beginning of each batch.

In [8]:
def process_decoding_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for dencoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    go = [tf.constant([target_vocab_to_int['<GO>'],]) for i in range(batch_size)]
    data = []
    for sentence in tf.unstack(target_data, batch_size):
        sentence = sentence[:-1]
        data.append(sentence)
    data = tf.stack(data)
    data = tf.concat([go,data],1)
    return data

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_decoding_input(process_decoding_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer using [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn).

In [9]:
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :return: RNN state
    """
    # TODO: Implement Function
    cell = tf.contrib.rnn.MultiRNNCell([tf.contrib.rnn.DropoutWrapper(tf.contrib.rnn.BasicLSTMCell(rnn_size),
                                                                     output_keep_prob=keep_prob) for _ in range(num_layers)])
    outputs, final_state = tf.nn.dynamic_rnn(cell, rnn_inputs, dtype=tf.float32) 
    return final_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create training logits using [`tf.contrib.seq2seq.simple_decoder_fn_train()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_train) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder).  Apply the `output_fn` to the [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) outputs.

In [10]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope,
                         output_fn, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param sequence_length: Sequence Length
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Train Logits
    """
    # TODO: Implement Function
    dynamic_fn_train =  tf.contrib.seq2seq.simple_decoder_fn_train(encoder_state)
    #cell = tf.contrib.rnn.DropoutWrapper(dec_cell,output_keep_prob=keep_prob)
    outputs_train,_,_ =  tf.contrib.seq2seq.dynamic_rnn_decoder(dec_cell,dynamic_fn_train, dec_embed_input,sequence_length)
    return output_fn(tf.nn.dropout(outputs_train,keep_prob))


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference logits using [`tf.contrib.seq2seq.simple_decoder_fn_inference()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_inference) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder). 

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id,
                         maximum_length, vocab_size, decoding_scope, output_fn, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param maximum_length: The maximum allowed time steps to decode
    :param vocab_size: Size of vocabulary
    :param decoding_scope: TensorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Inference Logits
    """
    # TODO: Implement Function
    dynamic_fn_inference =  tf.contrib.seq2seq.simple_decoder_fn_inference(output_fn, encoder_state, dec_embeddings, 
                                                                           start_of_sequence_id,end_of_sequence_id,
                                                                           maximum_length,vocab_size)
    #cell = tf.contrib.rnn.DropoutWrapper(dec_cell,output_keep_prob=keep_prob)
    outputs_inference, _ ,_ =  tf.contrib.seq2seq.dynamic_rnn_decoder(dec_cell,dynamic_fn_inference)
    return outputs_inference


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

- Create RNN cell for decoding using `rnn_size` and `num_layers`.
- Create the output fuction using [`lambda`](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions) to transform it's input, logits, to class logits.
- Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope, output_fn, keep_prob)` function to get the training logits.
- Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, maximum_length, vocab_size, decoding_scope, output_fn, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size,
                   num_layers, target_vocab_to_int, keep_prob):
    """
    Create decoding layer
    :param dec_embed_input: Decoder embedded input
    :param dec_embeddings: Decoder embeddings
    :param encoder_state: The encoded state
    :param vocab_size: Size of vocabulary
    :param sequence_length: Sequence Length
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # TODO: Implement Function
    start_of_sequence_id = target_vocab_to_int['<GO>']
    end_of_sequence_id = target_vocab_to_int['<EOS>']
    dec_cell = tf.contrib.rnn.MultiRNNCell([tf.contrib.rnn.DropoutWrapper(tf.contrib.rnn.BasicLSTMCell(rnn_size), 
                                                                         output_keep_prob=keep_prob) for _ in range(num_layers)])
    with tf.variable_scope("decoder") as decoding_scope:
        #output_fn = lambda x: tf.contrib.layers.linear(x, vocab_size,scope=decoding_scope)
        output_fn = lambda x: tf.contrib.layers.fully_connected(x, vocab_size,activation_fn=None,scope=decoding_scope)
        training_logits = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, 
                                           decoding_scope, output_fn, keep_prob) 
        decoding_scope.reuse_variables()
        inference_logits = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, 
                                            end_of_sequence_id, sequence_length, vocab_size, decoding_scope, 
                                            output_fn, keep_prob)
    
    return training_logits, inference_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)`.
- Process target data using your `process_decoding_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)`.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size, sequence_length, source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size, rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param sequence_length: Sequence Length
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # TODO: Implement Function
    enc_embeddings = tf.Variable(tf.random_uniform((source_vocab_size, enc_embedding_size), -1, 1))
    rnn_inputs = tf.nn.embedding_lookup(enc_embeddings, input_data)
    encoder_state = encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)
    decoding_input = process_decoding_input(target_data, target_vocab_to_int, batch_size) 
    dec_embeddings = tf.Variable(tf.random_uniform((target_vocab_size, dec_embedding_size), -1, 1))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, decoding_input)
    training_logits, inference_logits = decoding_layer(dec_embed_input, dec_embeddings, encoder_state, target_vocab_size, 
                                                       sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)
    return training_logits, inference_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability

In [14]:
# Number of Epochs
epochs = 3
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 512
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.75

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_source_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob = model_inputs()
    sequence_length = tf.placeholder_with_default(max_source_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)
    
    train_logits, inference_logits = seq2seq_model(
        tf.reverse(input_data, [-1]), targets, keep_prob, batch_size, sequence_length, len(source_vocab_to_int), len(target_vocab_to_int),
        encoding_embedding_size, decoding_embedding_size, rnn_size, num_layers, target_vocab_to_int)

    tf.identity(inference_logits, 'logits')
    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            train_logits,
            targets,
            tf.ones([input_shape[0], sequence_length]))

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import time

def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1]), (0,0)],
            'constant')

    return np.mean(np.equal(target, np.argmax(logits, 2)))

train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]

valid_source = helper.pad_sentence_batch(source_int_text[:batch_size])
valid_target = helper.pad_sentence_batch(target_int_text[:batch_size])

with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch) in enumerate(
                helper.batch_data(train_source, train_target, batch_size)):
            start_time = time.time()
            
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 sequence_length: target_batch.shape[1],
                 keep_prob: keep_probability})
            
            batch_train_logits = sess.run(
                inference_logits,
                {input_data: source_batch, keep_prob: 1.0})
            batch_valid_logits = sess.run(
                inference_logits,
                {input_data: valid_source, keep_prob: 1.0})
                
            train_acc = get_accuracy(target_batch, batch_train_logits)
            valid_acc = get_accuracy(np.array(valid_target), batch_valid_logits)
            end_time = time.time()
            print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.3f}, Validation Accuracy: {:>6.3f}, Loss: {:>6.3f}'
                  .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    0/269 - Train Accuracy:  0.273, Validation Accuracy:  0.339, Loss:  5.901
Epoch   0 Batch    1/269 - Train Accuracy:  0.260, Validation Accuracy:  0.336, Loss:  5.448
Epoch   0 Batch    2/269 - Train Accuracy:  0.296, Validation Accuracy:  0.339, Loss:  4.343
Epoch   0 Batch    3/269 - Train Accuracy:  0.307, Validation Accuracy:  0.367, Loss:  4.895
Epoch   0 Batch    4/269 - Train Accuracy:  0.241, Validation Accuracy:  0.308, Loss:  3.771
Epoch   0 Batch    5/269 - Train Accuracy:  0.310, Validation Accuracy:  0.378, Loss:  3.940
Epoch   0 Batch    6/269 - Train Accuracy:  0.346, Validation Accuracy:  0.375, Loss:  3.510
Epoch   0 Batch    7/269 - Train Accuracy:  0.329, Validation Accuracy:  0.362, Loss:  3.265
Epoch   0 Batch    8/269 - Train Accuracy:  0.290, Validation Accuracy:  0.356, Loss:  3.280
Epoch   0 Batch    9/269 - Train Accuracy:  0.325, Validation Accuracy:  0.365, Loss:  3.103
Epoch   0 Batch   10/269 - Train Accuracy:  0.316, Validation Accuracy

Epoch   0 Batch   89/269 - Train Accuracy:  0.546, Validation Accuracy:  0.548, Loss:  0.929
Epoch   0 Batch   90/269 - Train Accuracy:  0.526, Validation Accuracy:  0.573, Loss:  0.979
Epoch   0 Batch   91/269 - Train Accuracy:  0.539, Validation Accuracy:  0.558, Loss:  0.894
Epoch   0 Batch   92/269 - Train Accuracy:  0.560, Validation Accuracy:  0.568, Loss:  0.902
Epoch   0 Batch   93/269 - Train Accuracy:  0.554, Validation Accuracy:  0.555, Loss:  0.867
Epoch   0 Batch   94/269 - Train Accuracy:  0.551, Validation Accuracy:  0.562, Loss:  0.908
Epoch   0 Batch   95/269 - Train Accuracy:  0.585, Validation Accuracy:  0.587, Loss:  0.881
Epoch   0 Batch   96/269 - Train Accuracy:  0.590, Validation Accuracy:  0.592, Loss:  0.872
Epoch   0 Batch   97/269 - Train Accuracy:  0.556, Validation Accuracy:  0.574, Loss:  0.859
Epoch   0 Batch   98/269 - Train Accuracy:  0.578, Validation Accuracy:  0.578, Loss:  0.847
Epoch   0 Batch   99/269 - Train Accuracy:  0.567, Validation Accuracy

Epoch   0 Batch  178/269 - Train Accuracy:  0.737, Validation Accuracy:  0.740, Loss:  0.498
Epoch   0 Batch  179/269 - Train Accuracy:  0.738, Validation Accuracy:  0.742, Loss:  0.486
Epoch   0 Batch  180/269 - Train Accuracy:  0.745, Validation Accuracy:  0.748, Loss:  0.479
Epoch   0 Batch  181/269 - Train Accuracy:  0.741, Validation Accuracy:  0.743, Loss:  0.495
Epoch   0 Batch  182/269 - Train Accuracy:  0.748, Validation Accuracy:  0.742, Loss:  0.472
Epoch   0 Batch  183/269 - Train Accuracy:  0.793, Validation Accuracy:  0.750, Loss:  0.411
Epoch   0 Batch  184/269 - Train Accuracy:  0.747, Validation Accuracy:  0.759, Loss:  0.490
Epoch   0 Batch  185/269 - Train Accuracy:  0.765, Validation Accuracy:  0.762, Loss:  0.459
Epoch   0 Batch  186/269 - Train Accuracy:  0.750, Validation Accuracy:  0.759, Loss:  0.474
Epoch   0 Batch  187/269 - Train Accuracy:  0.750, Validation Accuracy:  0.760, Loss:  0.449
Epoch   0 Batch  188/269 - Train Accuracy:  0.758, Validation Accuracy

Epoch   0 Batch  267/269 - Train Accuracy:  0.852, Validation Accuracy:  0.854, Loss:  0.249
Epoch   1 Batch    0/269 - Train Accuracy:  0.877, Validation Accuracy:  0.863, Loss:  0.255
Epoch   1 Batch    1/269 - Train Accuracy:  0.867, Validation Accuracy:  0.871, Loss:  0.237
Epoch   1 Batch    2/269 - Train Accuracy:  0.862, Validation Accuracy:  0.868, Loss:  0.233
Epoch   1 Batch    3/269 - Train Accuracy:  0.870, Validation Accuracy:  0.862, Loss:  0.238
Epoch   1 Batch    4/269 - Train Accuracy:  0.864, Validation Accuracy:  0.863, Loss:  0.235
Epoch   1 Batch    5/269 - Train Accuracy:  0.879, Validation Accuracy:  0.861, Loss:  0.228
Epoch   1 Batch    6/269 - Train Accuracy:  0.872, Validation Accuracy:  0.851, Loss:  0.216
Epoch   1 Batch    7/269 - Train Accuracy:  0.875, Validation Accuracy:  0.861, Loss:  0.221
Epoch   1 Batch    8/269 - Train Accuracy:  0.876, Validation Accuracy:  0.864, Loss:  0.224
Epoch   1 Batch    9/269 - Train Accuracy:  0.875, Validation Accuracy

Epoch   1 Batch   88/269 - Train Accuracy:  0.923, Validation Accuracy:  0.929, Loss:  0.098
Epoch   1 Batch   89/269 - Train Accuracy:  0.930, Validation Accuracy:  0.916, Loss:  0.094
Epoch   1 Batch   90/269 - Train Accuracy:  0.925, Validation Accuracy:  0.915, Loss:  0.097
Epoch   1 Batch   91/269 - Train Accuracy:  0.931, Validation Accuracy:  0.934, Loss:  0.085
Epoch   1 Batch   92/269 - Train Accuracy:  0.946, Validation Accuracy:  0.925, Loss:  0.081
Epoch   1 Batch   93/269 - Train Accuracy:  0.928, Validation Accuracy:  0.926, Loss:  0.084
Epoch   1 Batch   94/269 - Train Accuracy:  0.924, Validation Accuracy:  0.934, Loss:  0.102
Epoch   1 Batch   95/269 - Train Accuracy:  0.936, Validation Accuracy:  0.938, Loss:  0.086
Epoch   1 Batch   96/269 - Train Accuracy:  0.908, Validation Accuracy:  0.935, Loss:  0.089
Epoch   1 Batch   97/269 - Train Accuracy:  0.928, Validation Accuracy:  0.928, Loss:  0.088
Epoch   1 Batch   98/269 - Train Accuracy:  0.925, Validation Accuracy

Epoch   1 Batch  177/269 - Train Accuracy:  0.949, Validation Accuracy:  0.949, Loss:  0.055
Epoch   1 Batch  178/269 - Train Accuracy:  0.949, Validation Accuracy:  0.953, Loss:  0.053
Epoch   1 Batch  179/269 - Train Accuracy:  0.938, Validation Accuracy:  0.949, Loss:  0.058
Epoch   1 Batch  180/269 - Train Accuracy:  0.939, Validation Accuracy:  0.949, Loss:  0.058
Epoch   1 Batch  181/269 - Train Accuracy:  0.937, Validation Accuracy:  0.952, Loss:  0.062
Epoch   1 Batch  182/269 - Train Accuracy:  0.947, Validation Accuracy:  0.955, Loss:  0.058
Epoch   1 Batch  183/269 - Train Accuracy:  0.953, Validation Accuracy:  0.953, Loss:  0.045
Epoch   1 Batch  184/269 - Train Accuracy:  0.944, Validation Accuracy:  0.954, Loss:  0.054
Epoch   1 Batch  185/269 - Train Accuracy:  0.953, Validation Accuracy:  0.951, Loss:  0.057
Epoch   1 Batch  186/269 - Train Accuracy:  0.954, Validation Accuracy:  0.955, Loss:  0.053
Epoch   1 Batch  187/269 - Train Accuracy:  0.950, Validation Accuracy

Epoch   1 Batch  266/269 - Train Accuracy:  0.959, Validation Accuracy:  0.961, Loss:  0.037
Epoch   1 Batch  267/269 - Train Accuracy:  0.956, Validation Accuracy:  0.961, Loss:  0.048
Epoch   2 Batch    0/269 - Train Accuracy:  0.965, Validation Accuracy:  0.958, Loss:  0.052
Epoch   2 Batch    1/269 - Train Accuracy:  0.957, Validation Accuracy:  0.959, Loss:  0.046
Epoch   2 Batch    2/269 - Train Accuracy:  0.954, Validation Accuracy:  0.958, Loss:  0.046
Epoch   2 Batch    3/269 - Train Accuracy:  0.957, Validation Accuracy:  0.962, Loss:  0.044
Epoch   2 Batch    4/269 - Train Accuracy:  0.948, Validation Accuracy:  0.960, Loss:  0.045
Epoch   2 Batch    5/269 - Train Accuracy:  0.960, Validation Accuracy:  0.957, Loss:  0.044
Epoch   2 Batch    6/269 - Train Accuracy:  0.964, Validation Accuracy:  0.959, Loss:  0.038
Epoch   2 Batch    7/269 - Train Accuracy:  0.957, Validation Accuracy:  0.957, Loss:  0.042
Epoch   2 Batch    8/269 - Train Accuracy:  0.963, Validation Accuracy

Epoch   2 Batch   87/269 - Train Accuracy:  0.959, Validation Accuracy:  0.952, Loss:  0.038
Epoch   2 Batch   88/269 - Train Accuracy:  0.957, Validation Accuracy:  0.961, Loss:  0.040
Epoch   2 Batch   89/269 - Train Accuracy:  0.967, Validation Accuracy:  0.962, Loss:  0.034
Epoch   2 Batch   90/269 - Train Accuracy:  0.958, Validation Accuracy:  0.956, Loss:  0.035
Epoch   2 Batch   91/269 - Train Accuracy:  0.965, Validation Accuracy:  0.963, Loss:  0.033
Epoch   2 Batch   92/269 - Train Accuracy:  0.967, Validation Accuracy:  0.961, Loss:  0.032
Epoch   2 Batch   93/269 - Train Accuracy:  0.962, Validation Accuracy:  0.957, Loss:  0.035
Epoch   2 Batch   94/269 - Train Accuracy:  0.949, Validation Accuracy:  0.953, Loss:  0.045
Epoch   2 Batch   95/269 - Train Accuracy:  0.965, Validation Accuracy:  0.962, Loss:  0.031
Epoch   2 Batch   96/269 - Train Accuracy:  0.953, Validation Accuracy:  0.966, Loss:  0.037
Epoch   2 Batch   97/269 - Train Accuracy:  0.954, Validation Accuracy

Epoch   2 Batch  176/269 - Train Accuracy:  0.968, Validation Accuracy:  0.968, Loss:  0.035
Epoch   2 Batch  177/269 - Train Accuracy:  0.970, Validation Accuracy:  0.972, Loss:  0.029
Epoch   2 Batch  178/269 - Train Accuracy:  0.969, Validation Accuracy:  0.969, Loss:  0.027
Epoch   2 Batch  179/269 - Train Accuracy:  0.965, Validation Accuracy:  0.972, Loss:  0.031
Epoch   2 Batch  180/269 - Train Accuracy:  0.972, Validation Accuracy:  0.972, Loss:  0.029
Epoch   2 Batch  181/269 - Train Accuracy:  0.962, Validation Accuracy:  0.969, Loss:  0.035
Epoch   2 Batch  182/269 - Train Accuracy:  0.964, Validation Accuracy:  0.970, Loss:  0.030
Epoch   2 Batch  183/269 - Train Accuracy:  0.972, Validation Accuracy:  0.968, Loss:  0.024
Epoch   2 Batch  184/269 - Train Accuracy:  0.970, Validation Accuracy:  0.964, Loss:  0.027
Epoch   2 Batch  185/269 - Train Accuracy:  0.975, Validation Accuracy:  0.965, Loss:  0.031
Epoch   2 Batch  186/269 - Train Accuracy:  0.968, Validation Accuracy

Epoch   2 Batch  265/269 - Train Accuracy:  0.972, Validation Accuracy:  0.970, Loss:  0.024
Epoch   2 Batch  266/269 - Train Accuracy:  0.969, Validation Accuracy:  0.972, Loss:  0.021
Epoch   2 Batch  267/269 - Train Accuracy:  0.972, Validation Accuracy:  0.974, Loss:  0.029
Model Trained and Saved


### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [19]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    seq = []
    for word in sentence.split():
        try:
            seq.append(vocab_to_int[word])
        except:
            seq.append(vocab_to_int['<UNK>'])
    return seq


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [20]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('logits:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence], keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in np.argmax(translate_logits, 1)]))
print('  French Words: {}'.format([target_int_to_vocab[i] for i in np.argmax(translate_logits, 1)]))

Input
  Word Ids:      [64, 164, 230, 25, 148, 124, 16]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [60, 97, 238, 87, 168, 254, 323, 178, 1]
  French Words: ['il', 'a', 'vu', 'un', 'vieux', 'camion', 'jaune', '.', '<EOS>']


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.